src/hardware/dac.rs

```diff
@@ -0,0 +1,279 @@
+                pins.max_i_neg2_pin,
+                pins.max_i_neg3_pin,
+            ),
+            F_PWM.khz(),
+            tim_rcc.1,
+            clocks,
+        );
+        init_pwm_pin(&mut max_i_neg0);
+        init_pwm_pin(&mut max_i_neg1);
+        init_pwm_pin(&mut max_i_neg2);
+        init_pwm_pin(&mut max_i_neg3);
+
+        Pwm {
+            max_v0,
+            max_v1,
+            max_v2,
+            max_v3,
+            max_i_pos0,
+            max_i_pos1,
+            max_i_pos2,
+            max_i_pos3,
+            max_i_neg0,
+            max_i_neg1,
+            max_i_neg2,
+            max_i_neg3,
+        }
+    }
+
+    /// Set PWM to limit current.
+    pub fn set(&mut self, ch: Channel, lim: Limit, val: f32) {
+        fn set_pwm<P: PwmPin<Duty = u16>>(pin: &mut P, duty: f32) {
+            let max = pin.get_max_duty();
+            let value = ((duty * (max as f32)) as u16).min(max);
+            pin.set_duty(value);
+        }
+        match (ch, lim) {
+            (Channel::Ch0, Limit::MaxV) => set_pwm(&mut self.max_v0, v_to_pwm(val)),
+            (Channel::Ch1, Limit::MaxV) => set_pwm(&mut self.max_v1, v_to_pwm(val)),
+            (Channel::Ch2, Limit::MaxV) => set_pwm(&mut self.max_v2, v_to_pwm(val)),
+            (Channel::Ch3, Limit::MaxV) => set_pwm(&mut self.max_v3, v_to_pwm(val)),
+            (Channel::Ch0, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos0, i_to_pwm(val)),
+            (Channel::Ch1, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos1, i_to_pwm(val)),
+            (Channel::Ch2, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos2, i_to_pwm(val)),
+            (Channel::Ch3, Limit::MaxIPos) => set_pwm(&mut self.max_i_pos3, i_to_pwm(val)),
+            (Channel::Ch0, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg0, i_to_pwm(val)),
+            (Channel::Ch1, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg1, i_to_pwm(val)),
+            (Channel::Ch2, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg2, i_to_pwm(val)),
+            (Channel::Ch3, Limit::MaxINeg) => set_pwm(&mut self.max_i_neg3, i_to_pwm(val)),
+        }
+    }
+}
+
+pub struct DacGpio {
+    pub sync0: PG3<Output<PushPull>>,
+    pub sync1: PG2<Output<PushPull>>,
+    pub sync2: PG1<Output<PushPull>>,
+    pub sync3: PG0<Output<PushPull>>,
+    pub shdn0: PG4<Output<PushPull>>,
+    pub shdn1: PG5<Output<PushPull>>,
+    pub shdn2: PG6<Output<PushPull>>,
+    pub shdn3: PG7<Output<PushPull>>,
+}
+
+/// DAC: https://www.analog.com/media/en/technical-documentation/data-sheets/AD5680.pdf
+/// Peltier Driver: https://datasheets.maximintegrated.com/en/ds/MAX1968-MAX1969.pdf
+pub struct Dac {
+    spi: Spi<SPI3, Enabled, u8>,
+    pub val: [u32; 4],
+    gpio: DacGpio,
+}
+
+impl Dac {
+    pub fn new(
+        clocks: &CoreClocks,
+        prec: rec::Spi3,
+        spi3: SPI3,
+        sck: PC10<Alternate<AF6>>,
+        mosi: PC12<Alternate<AF6>>,
+        gpio: DacGpio,
+    ) -> Self {
+        let spi = spi3.spi(
```

Comment:
Ah. That would make the hypothetical crate even less useful. It would only take over the job of doing `self.bus.write(&mut value.to_be_bytes()[1..])`.

Reply:
I would surely need another layer to embed the IOs that have to do with the DAC. I don't want to do that separately main/setup.